# Lecture 3: Data Retrieval

:::{admonition} Learning Objectives
:class: tip
After this lecture, you will be able to:
- Explain why efficient data retrieval matters at scale
- Choose appropriate data formats (CSV vs Parquet) and data types
- Write performant queries using Polars and DuckDB
- Understand when to use in-memory processing vs databases
- Interact with relational databases from Python
- Navigate the data storage and management landscape {cite}`kleppmann2017designing`
- Explain the difference between row-oriented and columnar storage formats
- Recognise when a dataset has outgrown pandas and identify the right next tool
:::

This lecture builds on the pandas material from [Lecture 2](lecture_2.ipynb) and prepares the ground for [Lecture 4](lecture_4.ipynb), where we move from data engineering to visualisation and exploratory data analysis. The storage and database concepts here follow the systems perspective of {cite}`kleppmann2017designing`; the "know your regime" advice draws on {cite}`huyen2022designing`.

## Table of Contents

- [Motivation: working with (biggish) data](#motivation-working-with-biggish-data)
- [How to speed things up](#how-to-speed-things-up)
- [Python and databases](#python-and-databases)
- [Data storage & management landscape](#data-storage-management-landscape)
- [Scale awareness: when pandas is not enough](#scale-awareness-when-pandas-is-not-enough)
- [Repository structure: .gitignore, pre-commit, type hints](#repository-structure-gitignore-pre-commit-type-hints)

A [Key Takeaways](#key-takeaways) box and an [Exercises](#exercises) section close the lecture.

:::{note}
This notebook imports from the `fun_ds` package. To run the cells, clone the
repository and follow the setup instructions on the [home page](/intro).
:::

In [1]:
import pathlib
import tempfile

import numpy as np
import pandas as pd

from fun_ds.data import load_california_housing
from fun_ds.plotting import set_lecture_style

set_lecture_style()
pd.set_option("display.max_columns", 50)

df = load_california_housing()
print(df.shape)
df.head()

(20640, 9)


,MedInc,HouseAge,AveRooms,AveBedrms,Population,AveOccup,Latitude,Longitude,MedHouseVal
0,8.3252,41.0,6.984127,1.023810,322.0,2.555556,37.88,-122.23,4.526
1,8.3014,21.0,6.238137,0.971880,2401.0,2.109842,37.86,-122.22,3.585
2,7.2574,52.0,8.288136,1.073446,496.0,2.802260,37.85,-122.24,3.521
3,5.6431,52.0,5.817352,1.073059,558.0,2.547945,37.85,-122.25,3.413
4,3.8462,52.0,6.281853,1.081081,565.0,2.181467,37.85,-122.25,3.422


## Motivation: Working with (Biggish) Data

Data science is, before anything else, a *data-first* discipline. Models — however sophisticated — inherit the quality, coverage, and structure of the data they are fitted on, and the practitioner's day-to-day time budget reflects that. A widely cited figure among practitioners is that data scientists spend roughly **60–80% of their time** on data collection, cleaning, and preparation rather than on modelling itself {cite}`kuhn2019feature,huyen2022designing`. Whether the exact number is 60 or 80, the qualitative point is uncontroversial: fluency with retrieval and storage is not a peripheral skill but the *central* one.

If you do not use the right tools, you will spend most of your time **waiting** on data.

Two motivating case studies:
- Automating a business process for an insurance company (>20 tables, >500M rows, >300 columns)
- Training an LLM (Llama 3.1: 15 trillion tokens of pre-training data)

$\rightarrow$ Being able to handle large amounts of data efficiently is a key skill, and — as {cite}`huyen2022designing` argues for production ML — often the difference between a system that ships and one that stalls in a notebook.

## How to Speed Things Up

In the lecture we load >178 million rows of NYC taxi data in under two seconds with Polars and DuckDB. That dataset is too large to ship with this book, so the cells below demonstrate the same principles on (a scaled-up copy of) California Housing — every cell runs on a laptop in seconds. The key principles:
1. Only use what you need
2. Choose the right data types
3. Optimise your queries
4. Understand in-memory vs on-disk trade-offs
5. Use appropriate database management systems

### Why You Should Not Work with CSV

:::{warning}
CSV files are human-readable but problematic at scale:
- No type information (everything is text)
- No compression
- Must read entire file to query a subset
- Parsing is slow and error-prone
:::

(row-vs-columnar)=
#### Row- vs. column-oriented storage

The deeper reason CSV struggles at analytical scale is that it is **row-oriented**: bytes on disk follow the order *(row 1, all columns) → (row 2, all columns) → …*. This suits workloads where each operation touches a few complete records — think of an insurance claims system inserting or fetching one claim at a time. In contrast, **columnar** formats such as Apache Parquet lay bytes out in *column chunks*: all values of column A, then all values of column B, and so on. This suits analytical workloads — aggregations, filters, and joins over a handful of columns across many rows {cite}`kleppmann2017designing`. (The database world calls these two workload types OLTP and OLAP; see [OLTP vs OLAP](#oltp-vs-olap) below.)

| Format | Storage order | Read a full row | Read one column | Write one row |
|---|---|---|---|---|
| CSV | Row-oriented | Fast | Slow (scan all) | Append is easy |
| Parquet | Column-oriented | Slow (reconstruct) | **Very fast** | Requires rewrite |
| Arrow (in-memory) | Column-oriented | Slow | **Very fast** | Requires rewrite |

Three properties follow from the columnar layout, and together they explain almost every reported performance advantage of Parquet over CSV under typical analytical workloads:

1. **Higher compressibility.** Values *within* a column chunk tend to be similar in type and often in magnitude (e.g. all `float64` incomes, all short country codes). Encoding schemes such as run-length encoding, dictionary encoding, and bit-packing exploit this local homogeneity and routinely deliver 5–10× on-disk compression. A row-oriented layout mixes types byte-by-byte and cannot compress as effectively.
2. **Less I/O per query.** Analytical queries usually touch a small subset of columns (`SELECT MedInc, MedHouseVal ...`). With a column layout, the query engine can perform *projection pushdown* and read only the relevant column chunks from disk. With CSV every query pays for reading every column.
3. **Vectorised execution.** Once in memory, a contiguous column of one type is exactly what SIMD CPU instructions operate on, so engines process a whole batch of values per instruction.

:::{note}
These properties are qualitative; the exact speed-up depends on the compression codec, the query, the storage medium, and the CPU. Claims like "Parquet is 10× faster" should therefore always be read as "under typical analytical workloads, on this dataset, with this engine".
:::

In [2]:
# CSV vs Parquet: file size and load time on ~100k rows of California Housing.
# We repeat the housing frame to reach ~100k rows and write to a temp dir,
# so nothing is left behind in the repository.
# Caveat: repeated rows compress unusually well, so treat the size ratio as indicative.
import timeit

rows = pd.concat([df] * 5, ignore_index=True)  # ~103k rows
print(f"benchmark frame shape: {rows.shape}")


def best_of(fn, repeat=5):
    """Best wall-clock time (s) over `repeat` runs — less noisy than a single run."""
    return min(timeit.repeat(fn, number=1, repeat=repeat))


with tempfile.TemporaryDirectory() as tmp:
    csv_path = pathlib.Path(tmp) / "housing.csv"
    parquet_path = pathlib.Path(tmp) / "housing.parquet"

    rows.to_csv(csv_path, index=False)
    try:
        rows.to_parquet(parquet_path, index=False)
        parquet_available = True
    except ImportError:
        parquet_available = False
        print(
            "pyarrow/fastparquet not installed; it is part of the course environment: run `pixi install`."
        )

    csv_size = csv_path.stat().st_size / 1e6
    print(f"CSV file size:     {csv_size:6.2f} MB")

    # The first call to a reader pays one-off costs (lazy imports, allocations).
    # Warm up once so we compare steady-state read speed, not start-up time.
    pd.read_csv(csv_path)
    csv_load = best_of(lambda: pd.read_csv(csv_path))
    print(f"CSV load time:     {csv_load:6.3f} s")

    if parquet_available:
        parquet_size = parquet_path.stat().st_size / 1e6
        print(
            f"Parquet file size: {parquet_size:6.2f} MB  ({csv_size / parquet_size:.1f}x smaller)"
        )

        pd.read_parquet(parquet_path)  # warm-up (imports pyarrow on first use)
        parquet_load = best_of(lambda: pd.read_parquet(parquet_path))
        print(
            f"Parquet load time: {parquet_load:6.3f} s  ({csv_load / parquet_load:.1f}x faster)"
        )

benchmark frame shape: (103200, 9)


CSV file size:       9.48 MB
CSV load time:      0.031 s
Parquet file size:   1.91 MB  (5.0x smaller)


Parquet load time:  0.001 s  (26.1x faster)


**Expected output.** On typical hardware, the Parquet file is roughly 5–10× smaller than the CSV and loads many times faster (often 10× or more) because it stores each column contiguously in a binary, typed, compressed form — there is no text to parse. The gap widens as datasets grow.

:::{tip}
Notice the warm-up call before timing. If you time a single, first `pd.read_parquet` call you mostly measure the one-off cost of importing `pyarrow`, and Parquet can even look *slower* than CSV. Benchmark hygiene — warm up, repeat, take the best (or median) — matters as much as the format choice.
:::

#### Schema evolution

Real datasets rarely have a fixed schema over their lifetime: business logic changes, new features are collected, deprecated fields are dropped. **Schema evolution** is the study of how storage formats accommodate these changes without full rewrites.

Parquet handles schema evolution *asymmetrically*:

- **Adding a column is cheap.** New files are written with the extra column; older files without it are read as `NULL` for that column. This is because Parquet stores each column as an independent chunk with its own metadata.
- **Removing a column is cheap.** Readers simply project the columns they still need.
- **Renaming a column is *not* cheap in plain Parquet.** Column names are written into every file's footer, so a rename means rewriting the files (or maintaining a name mapping in your reader), and readers that match on name treat old and new files as having different columns. Metadata-only renames need a *table format* layered on top of Parquet, such as [Apache Iceberg](https://iceberg.apache.org/) or [Delta Lake](https://delta.io/), which tracks columns by ID rather than by name.
- **Changing a column's type is expensive.** Column chunks are physically encoded for a specific logical type; converting `int32` → `int64` or `string` → `categorical` requires rewriting the affected files.

The take-away for pipeline design: prefer **additive** schema changes; treat type changes as a versioned migration, not a silent edit {cite}`kleppmann2017designing`.

:::{note}
**Apache Arrow — the in-memory columnar standard.** Parquet is a *disk* format. Its in-memory sibling is [Apache Arrow](https://arrow.apache.org/), a language-independent columnar memory layout that lets tools exchange tables cheaply, often without copying. Polars stores its columns in Arrow format natively; pandas uses Arrow for its default `str` dtype since pandas 3.0 and can use it for every column with `dtype_backend="pyarrow"`. DuckDB uses its own internal vector format, but it can *scan* Arrow data in place (zero-copy) and exports results to Arrow with a fast conversion. This shared interchange layout is what makes it cheap to move a table from a DuckDB query into a Polars pipeline, instead of paying a text-serialisation cost at every hop. The last hop into scikit-learn usually *does* copy: estimators want a single two-dimensional NumPy array of one numeric dtype, so `to_numpy()` assembles a new matrix from the separate columns.
:::

### Only Use What You Need

Column projection and row filtering at read time dramatically reduce I/O. Even with CSV you should tell pandas what you need up front — `pd.read_csv(path, usecols=[...])` to parse only some columns, and `pd.read_csv(path, nrows=10)` to peek at an unfamiliar file before committing to a full load — but CSV still has to *scan* every byte. A columnar format lets the reader skip the unneeded columns on disk entirely:

In [3]:
# Column projection: read only the columns you need from Parquet.

with tempfile.TemporaryDirectory() as tmp:
    parquet_path = pathlib.Path(tmp) / "housing.parquet"

    try:
        df.to_parquet(parquet_path, index=False)
    except ImportError:
        print("pyarrow is part of the course environment: run `pixi install`.")
    else:
        # Same benchmark hygiene as above: warm up, then take the best of several runs.
        full = pd.read_parquet(parquet_path)
        full_time = best_of(lambda: pd.read_parquet(parquet_path))
        print(
            f"all {full.shape[1]} columns:  {full.memory_usage(deep=True).sum() / 1e6:6.2f} MB in-memory,"
            f"  {full_time * 1000:6.1f} ms to load"
        )

        subset = pd.read_parquet(parquet_path, columns=["MedInc", "MedHouseVal"])
        subset_time = best_of(
            lambda: pd.read_parquet(parquet_path, columns=["MedInc", "MedHouseVal"])
        )
        print(
            f"only 2 columns:   {subset.memory_usage(deep=True).sum() / 1e6:6.2f} MB in-memory,"
            f"  {subset_time * 1000:6.1f} ms to load"
        )

        subset.head()

all 9 columns:    1.49 MB in-memory,     0.8 ms to load
only 2 columns:     0.33 MB in-memory,     0.5 ms to load


**Expected output.** The projected read loads a fraction of the bytes and materialises a smaller frame in memory. In row-oriented CSV the same operation would still read the whole file — projection is only useful once the format itself is columnar.

### Choose the Right Data Types

Using appropriate data types reduces memory usage and speeds up computation:
- `int64` vs `int32` vs `int16` — most counters do not need 64 bits
- `float64` vs `float32` — half the memory for most ML features
- Categorical vs string for repeated values (`pd.Categorical` / `pl.Categorical`)
- Native date/datetime types vs string parsing at every access

In [4]:
# Memory savings from right-sizing dtypes on California Housing.

# Add a low-cardinality string column to demonstrate the categorical case.
sample = df.copy()
sample["region"] = pd.cut(
    sample["Latitude"],
    bins=[-np.inf, 34, 36, 38, np.inf],
    labels=["SoCal", "Central-S", "Central-N", "NorCal"],
).astype(str)

before = sample.memory_usage(deep=True)
print("Before (default dtypes):")
print(before)
print(f"total: {before.sum() / 1e6:.2f} MB\n")

optimised = sample.copy()
float_cols = optimised.select_dtypes("float64").columns
# float32 keeps ~7 significant digits — ample for these features, but check before
# downcasting identifiers, money, or anything that needs exact arithmetic.
optimised[float_cols] = optimised[float_cols].astype("float32")
optimised["region"] = optimised["region"].astype("category")

after = optimised.memory_usage(deep=True)
print("After (float32 + category):")
print(after)
print(f"total: {after.sum() / 1e6:.2f} MB  ({before.sum() / after.sum():.1f}x smaller)")

Before (default dtypes):
Index             132
MedInc         165120
HouseAge       165120
AveRooms       165120
AveBedrms      165120
Population     165120
AveOccup       165120
Latitude       165120
Longitude      165120
MedHouseVal    165120
region         316595
dtype: int64
total: 1.80 MB

After (float32 + category):
Index            132
MedInc         82560
HouseAge       82560
AveRooms       82560
AveBedrms      82560
Population     82560
AveOccup       82560
Latitude       82560
Longitude      82560
MedHouseVal    82560
region         20702
dtype: int64
total: 0.76 MB  (2.4x smaller)


**Expected output.** Casting the float columns from `float64` to `float32` exactly halves the numeric memory footprint, and converting the low-cardinality string column to `category` replaces one stored string per row with a one-byte integer code per row plus a dictionary of four labels — here roughly a 15× saving on that column. Overall memory drops by roughly 2–3×; most of the absolute saving comes from the nine float columns simply because there are more of them. (In pandas 3.0 the default `str` dtype is Arrow-backed and fairly compact; in older pandas each string was a separate Python object, so the `category` saving was even larger.)

### Optimise Your Queries

Lazy evaluation enables query optimisation:
- Predicate pushdown (filter early)
- Projection pushdown (select columns early)
- Common subexpression elimination

In the output below, compare the two plans. In the naive plan the `FILTER` and `SELECT` sit *above* a scan that reads all 9 columns. In the optimised plan they have been folded into the `Parquet SCAN` itself: `SELECTION: col("MedInc") > 5.0` is **predicate pushdown**, and `PROJECT 2/9 COLUMNS` is **projection pushdown** — Polars noticed that `AveRooms` is selected but never used, and does not read it at all. With pandas you would have to spot and apply these optimisations by hand.

In [5]:
# Polars lazy evaluation: build a query plan, inspect it, then execute.
# We scan a Parquet file (in a temp dir) rather than an in-memory frame, so the
# plan can show filters and column selections being pushed into the file read.
try:
    import polars as pl

    with tempfile.TemporaryDirectory() as tmp:
        parquet_path = pathlib.Path(tmp) / "housing.parquet"
        df.to_parquet(parquet_path, index=False)

        lazy = (
            # Nothing is read yet — scan_parquet only starts building a plan.
            pl.scan_parquet(parquet_path)
            .filter(pl.col("MedInc") > 5)
            # AveRooms is selected here but never used below.
            .select(["MedInc", "AveRooms", "MedHouseVal"])
            .with_columns(pl.col("MedInc").round(0).alias("MedInc_rounded"))
            .group_by("MedInc_rounded")
            .agg(pl.mean("MedHouseVal").alias("mean_value"))
            .sort("MedInc_rounded")
        )

        # Plans read bottom-up: the scan is at the bottom, the final sort at the top.
        print("Naive plan (what we wrote):\n")
        print(lazy.explain(optimized=False))
        print("\nOptimised plan (what Polars will run):\n")
        print(lazy.explain())
        print("\nResult:\n")
        result = lazy.collect()  # only now is the file actually read
        print(result)

        # Same plan, run by the streaming engine: the file is processed in batches,
        # which is what lets a query over a larger-than-RAM file complete.
        # Batches are summed in a different order, so floating-point means can differ
        # in the last digit: compare with a tolerance, not exact equality.
        from polars.testing import assert_frame_equal

        streamed = lazy.collect(engine="streaming")
        assert_frame_equal(streamed, result)
        print(
            "\nStreaming engine gives the same result (up to floating-point rounding)."
        )
except ImportError:
    print(
        "Polars not installed. It is part of the course environment: run `pixi install`."
    )

Naive plan (what we wrote):

SORT BY [col("MedInc_rounded")]
  AGGREGATE[maintain_order: false]
    [col("MedHouseVal").mean().alias("mean_value")] BY [col("MedInc_rounded")]
    FROM
     WITH_COLUMNS:
     [col("MedInc").round().alias("MedInc_rounded")] 
      SELECT [col("MedInc"), col("AveRooms"), col("MedHouseVal")]
        FILTER col("MedInc") > 5.0
        FROM
          Parquet SCAN [/var/folders/lb/dnjbzjwj3nl6m39jdth2311c0000gp/T/tmpm12k_y5f/housing.parquet]
          PROJECT */9 COLUMNS
          ESTIMATED ROWS: 20640

Optimised plan (what Polars will run):

SORT BY [col("MedInc_rounded")]
  AGGREGATE[maintain_order: false]
    [col("MedHouseVal").mean().alias("mean_value")] BY [col("MedInc_rounded")]
    FROM
    simple π 2/2 ["MedInc_rounded", ... 1 other column]
       WITH_COLUMNS:
       [col("MedInc").round().alias("MedInc_rounded")] 
        Parquet SCAN [/var/folders/lb/dnjbzjwj3nl6m39jdth2311c0000gp/T/tmpm12k_y5f/housing.parquet]
        PROJECT 2/9 COLUMNS
        

### In-Memory vs On-Disk

| Approach | Pros | Cons |
|----------|------|------|
| In-memory (pandas, eager Polars) | Fast for small-medium data | Limited by RAM |
| On-disk / streaming (DuckDB, SQLite, Polars streaming engine) | Handles data larger than RAM | Slightly more overhead |

### Relational Database Management Systems & SQL

Why databases {cite}`kleppmann2017designing`?
- **Data integrity** — constraints and transactions guarantee that concurrent writes leave the data in a consistent state
- **Concurrent access** — many readers and writers without race conditions
- **Query optimisation built-in** — the query planner picks indexes, join order, and execution strategy
- **Standardised query language (SQL)** — portable across engines

(oltp-vs-olap)=
#### OLTP vs OLAP

A single "database" abstraction hides two very different engineering targets. Kleppmann {cite}`kleppmann2017designing` draws the standard distinction:

| Property | **OLTP** (Online Transaction Processing) | **OLAP** (Online Analytical Processing) |
|---|---|---|
| Typical workload | Many small point reads/writes (fetch one customer, insert one order) | Few large scans (aggregate over millions of rows) |
| Row selectivity | Reads a handful of rows | Reads most rows, few columns |
| Storage layout | Row-oriented (see [above](#row-vs-columnar)) | Column-oriented |
| Latency target | Millisecond | Second to minute |
| Example engines | PostgreSQL, MySQL, SQLite | DuckDB, ClickHouse, BigQuery, Snowflake |

DuckDB is an *OLAP* engine — bulk reads, aggregations, wide tables. SQLite and PostgreSQL are *OLTP*-first. Modern warehouses (Snowflake, BigQuery) are OLAP built for the cloud. Choose the one matching your workload: a data-science project usually wants OLAP; an application backend usually wants OLTP.

In [6]:
# DuckDB: query a pandas DataFrame directly, without loading it into a database first.
try:
    import duckdb

    result = duckdb.sql(
        """
        SELECT
            CASE
                WHEN MedInc < 2 THEN '0-2'
                WHEN MedInc < 4 THEN '2-4'
                WHEN MedInc < 6 THEN '4-6'
                WHEN MedInc < 8 THEN '6-8'
                ELSE '8+'
            END AS income_bin,
            COUNT(*)               AS n,
            AVG(MedHouseVal)       AS mean_value,
            AVG(AveRooms)          AS mean_rooms
        FROM df
        GROUP BY 1
        ORDER BY 1
        """
    ).df()
    print(result)
except ImportError:
    print(
        "DuckDB not installed. It is part of the course environment: run `pixi install`."
    )

  income_bin      n  mean_value  mean_rooms
0        0-2   2439    1.125127    4.462049
1        2-4  10077    1.679210    5.002252
2        4-6   5751    2.443854    5.897238
3        6-8   1682    3.444522    6.836097
4         8+    691    4.603707    7.743266


## Python and Databases

Key libraries for database interaction:
- **DuckDB** — embedded analytical database (OLAP), excellent for data science. Query pandas / Parquet / CSV directly with SQL.
- **SQLAlchemy** — ORM and database toolkit for production systems {cite}`huyen2022designing`. Speaks Postgres, MySQL, SQLite, and more through the same API.
- **sqlite3** — the built-in Python module. Perfect for prototypes, tests, and single-file storage.

In [7]:
# SQLAlchemy: write a DataFrame to an in-memory SQLite database and query it back.
try:
    from sqlalchemy import create_engine, text

    engine = create_engine("sqlite:///:memory:")
    df[["MedInc", "AveRooms", "MedHouseVal"]].to_sql(
        "housing", engine, index=False, if_exists="replace"
    )

    with engine.connect() as conn:
        result_rows = conn.execute(
            text(
                """
                SELECT ROUND(MedInc) AS income, AVG(MedHouseVal) AS mean_value, COUNT(*) AS n
                FROM housing
                GROUP BY ROUND(MedInc)
                ORDER BY income
                LIMIT 5
                """
            )
        ).fetchall()
    for row in result_rows:
        print(row)
except ImportError:
    print(
        "SQLAlchemy not installed. It is part of the course environment: run `pixi install`."
    )

(0.0, 1.6360849999999998, 12)
(1.0, 1.11308100871731, 803)
(2.0, 1.2325818096920746, 3962)
(3.0, 1.6887491440933033, 5316)
(4.0, 2.1061034962406016, 4522)


### Indexes: making point lookups cheap

The query planner can only "pick indexes" if an index exists. An **index** is a sorted auxiliary structure (a B-tree in SQLite and PostgreSQL) that maps values of a column to the rows holding them, so a selective filter can jump to the matching rows instead of scanning the whole table. The cell below runs the same narrow range query before and after `CREATE INDEX`, and asks SQLite how it plans to execute it (`EXPLAIN QUERY PLAN`).

**When would I use this?** Index the columns you *repeatedly filter or join on* when those filters are *selective* (they return a small fraction of rows) — the typical OLTP pattern. Indexes are not free: every insert and update must also maintain them, and they take disk space. For analytical scans that read most rows an index does not help; OLAP engines such as DuckDB instead keep min/max statistics per block of rows (*zone maps*) automatically, so you rarely create indexes there.

In [8]:
# Indexes in SQLite (built into Python): query plan and timing before and after.
import sqlite3

con = sqlite3.connect(":memory:")
rows.to_sql("housing", con, index=False)  # the ~103k-row frame from above

query = """
    SELECT COUNT(*), AVG(MedHouseVal)
    FROM housing
    WHERE MedInc BETWEEN 8.00 AND 8.05
"""


def plan(sql):
    """SQLite's description of how it will run `sql` (last column of the plan)."""
    return " | ".join(row[-1] for row in con.execute("EXPLAIN QUERY PLAN " + sql))


print("before:", plan(query))
t_scan = best_of(lambda: con.execute(query).fetchall(), repeat=20)

con.execute("CREATE INDEX idx_medinc ON housing (MedInc)")
print("after: ", plan(query))
t_index = best_of(lambda: con.execute(query).fetchall(), repeat=20)

print(f"\nfull scan: {t_scan * 1000:6.2f} ms   with index: {t_index * 1000:6.2f} ms")
print("result:", con.execute(query).fetchone())
con.close()

before: SCAN housing
after:  SEARCH housing USING INDEX idx_medinc (MedInc>? AND MedInc<?)

full scan:   1.47 ms   with index:   0.02 ms
result: (120, 4.244003333333334)


## Data Storage & Management Landscape

Overview of storage options:

| Format/System | Type | Best For |
|--------------|------|----------|
| CSV | File | Small data, interop |
| Parquet | File | Columnar analytics |
| SQLite | Embedded DB | Local apps, prototyping |
| DuckDB | Embedded DB | Analytics, data science |
| PostgreSQL | Server DB | Production systems |
| Delta Lake | Table format | Versioned data lakes |

## Scale Awareness: When Pandas Is Not Enough

### Columnar storage at scale

The [row- vs. column-oriented storage](#row-vs-columnar) comparison and the [OLTP vs OLAP](#oltp-vs-olap) split above are what decide how far a single machine gets you. The gap grows with the data:

:::{note}
A real-world example: a 1 GB CSV of 10M rows with 30 columns, where an analysis only uses 3 columns, requires reading all 1 GB from disk. The equivalent Parquet file may be 120 MB total and only ~15 MB of that needs to be read for those 3 columns. The exact figures vary with the codec, the query, and the hardware.
:::

### Lazy evaluation in Polars

Polars has two modes — **eager** (execute immediately) and **lazy** (build a query plan, optimise, then execute). The lazy API is what makes Polars competitive with on-disk databases for medium-sized data: `scan_parquet` + `collect` lets Polars decide how little of the file to read. The runnable demo in [Optimise Your Queries](#optimise-your-queries) shows the plan for a filter + project + group-by pipeline, with `filter` and `select` pushed down into the scan.

Lazy mode on its own does **not** make larger-than-RAM data work. A plain `collect()` still runs the plan on the default in-memory engine, which may need the (filtered, projected) data and intermediate results to fit in RAM. What makes it work is the **streaming engine**: `collect(engine="streaming")` processes the file in batches, and `sink_parquet(...)` (or `sink_csv`, …) writes the result straight to disk without ever materialising it in RAM. Laziness is the prerequisite — the engine can only stream a plan it can see in full — but streaming is the step that keeps memory bounded.

### SQL vs Python: where should the work happen?

A recurring question is: given that both DuckDB/Postgres *and* pandas/Polars can filter, join, and aggregate, where should each step of a pipeline live?

A useful rule of thumb — consistent with the systems perspective in {cite}`kleppmann2017designing` — is:

> **Filter and aggregate in SQL; analyse in Python.**

The reasoning is mechanical, not aesthetic:

- **SQL engines have a query optimiser.** Given a declarative query, DuckDB or Postgres will pick the join order, push filters down to the scan, and use indexes or zone maps to skip data. A hand-written pandas pipeline does none of this automatically.
- **Data movement is the dominant cost.** Every row that crosses the process/network boundary from the database into Python is paid for in serialisation and memory. Filtering and aggregating server-side collapses billions of rows into thousands *before* they ever hit Python.
- **Python is unmatched for what comes next.** Once the data is small enough to fit comfortably in memory, scikit-learn, statsmodels, matplotlib, and PyTorch have no serious SQL equivalents.

The concrete workflow: push filters, joins, and group-by aggregations into SQL; return a compact result frame; do the modelling, plotting, and interpretation in Python.

### Knowing when you have outgrown pandas

| Signal | What it means | What to reach for |
|---|---|---|
| `MemoryError` on load | Data does not fit in RAM | Polars lazy + streaming engine / DuckDB |
| Load takes > 30 s | CSV at scale | Convert to Parquet first |
| A join on two large DataFrames is slow | pandas joins are single-threaded | Polars (multi-threaded) or DuckDB |
| Data is partitioned by date/region | Hive-partitioned Parquet | `scan_parquet(..., hive_partitioning=True)` |
| Data > 100 GB on a single machine | Single-node limit | Spark / distributed systems |

:::{tip}
The boundary is roughly: **< 1 GB — pandas is fine; 1–100 GB — Polars (lazy, streaming) or DuckDB; > 100 GB — Spark or a cloud data warehouse**.
The important skill is recognising which regime you are in *before* the compute bill or the memory error tells you {cite}`huyen2022designing`.
:::

## Repository Structure: .gitignore, Pre-commit, Type Hints

As our projects grow, we need better tooling:
- **`.gitignore`** — keep large data files, model artefacts, and secrets out of version control (see [Lecture 2](lecture_2.ipynb) for details)
- **Pre-commit hooks** — run formatters (`ruff format`), linters (`ruff check`), and type checkers (`mypy`) automatically on every commit
- **Type hints** — document expected types for better IDE support, self-documenting code, and static error checking

In [9]:
# Type hints make function signatures precise and IDE-friendly.


def load_and_summarise(
    path: pathlib.Path,
    columns: list[str] | None = None,
) -> pd.DataFrame:
    """Load a Parquet file (with optional column projection) and print a compact summary.

    Parameters
    ----------
    path
        Path to the Parquet file.
    columns
        Columns to read. ``None`` reads all columns.

    Returns
    -------
    pd.DataFrame
        The loaded (and possibly projected) frame.
    """
    frame = pd.read_parquet(path, columns=columns)
    print(f"loaded {frame.shape[0]:,} rows x {frame.shape[1]} cols from {path.name}")
    print(f"memory: {frame.memory_usage(deep=True).sum() / 1e6:.2f} MB")
    return frame


# Demonstrate on the housing frame, written to a temp file so the repo stays clean.
with tempfile.TemporaryDirectory() as tmp:
    demo_path = pathlib.Path(tmp) / "demo.parquet"
    try:
        df.to_parquet(demo_path, index=False)
        out = load_and_summarise(demo_path, columns=["MedInc", "MedHouseVal"])
        print(out.head())
    except ImportError:
        print("Install `pyarrow` to run the round-trip.")

loaded 20,640 rows x 2 cols from demo.parquet
memory: 0.33 MB
   MedInc  MedHouseVal
0  8.3252        4.526
1  8.3014        3.585
2  7.2574        3.521
3  5.6431        3.413
4  3.8462        3.422


(key-takeaways)=
## Key Takeaways

:::{admonition} Key Takeaways
:class: important
- Choose Parquet over CSV for any non-trivial dataset — smaller on disk, faster to load, cheaper to query column-wise.
- Right-size your data types (`float32`, `int32`, `category`) — memory savings compound into runtime savings.
- Use lazy evaluation (Polars `scan_*` + `collect`) so filters and projections are pushed down into the file read; add the streaming engine (`collect(engine="streaming")`, `sink_*`) for larger-than-RAM data.
- Columnar formats (Parquet, Arrow) exist because analytics reads columns, not rows — OLAP and OLTP are different worlds {cite}`kleppmann2017designing`.
- Pick the right tool: pandas < 1 GB, Polars/DuckDB 1–100 GB, Spark or a warehouse > 100 GB.
- Databases give you integrity, concurrency, and a query optimiser you did not have to write.
- Keep large data out of Git; combine `.gitignore`, pre-commit, and type hints to keep production-ready code hygiene.
:::

(exercises)=
## Exercises

*Optional.* Each exercise trains one skill you will need whenever you take a real dataset from raw data to a model you can defend. They take 30–60 minutes each and end with **Done when** criteria you can check yourself; hints are collapsed so you can try first. The tutorials remain the core practice.

:::{admonition} Exercise 3.1 — What survives a round trip?
:class: exercise
**Skill.** Saving a cleaned dataset so that it reloads *exactly* as you left it: types, missing values and all.

```python
df = pd.DataFrame({
    "id": ["00123", "00456", "10789"],
    "region": pd.Categorical(["North", "South", "North"]),
    "rooms": pd.array([3, None, 5], dtype="Int64"),
    "listed": pd.to_datetime(["2024-01-05", "2024-02-11", "2024-03-20"]).tz_localize("Europe/London"),
    "is_new": [True, False, True],
})
```

1. Write `df` to CSV and to Parquet in a temporary directory, then read both back.
2. Write a small helper `dtype_diff(before, after) -> pd.DataFrame` that lists each column whose dtype changed, with old and new dtype.
3. For each column CSV damages, say what would break downstream (think of joins on `id` and one-hot encoding of `region`).
4. Compare the in-memory footprint (`memory_usage(deep=True)`) of `region` stored as `category` vs as strings on a 100 000-row version of the frame.

**Done when.**
- `pd.testing.assert_frame_equal(df, pd.read_parquet(path))` passes.
- `dtype_diff` on the CSV round trip returns every damaged column, and you can explain each one.
:::

:::{dropdown} Hint
CSV stores only text, so every type is re-inferred on reading. The leading zeros in `id` are the classic casualty. `read_csv(dtype=..., parse_dates=...)` can repair some damage, but you then have to maintain that schema by hand, which is exactly what Parquet stores for you.
:::

:::{admonition} Exercise 3.2 — Query Parquet without loading it
:class: exercise
**Skill.** Pushing filtering and aggregation to the storage engine instead of loading everything into pandas first.

Write the California Housing frame to Parquet. Using DuckDB (`duckdb.sql(...)`), write a single query **on the Parquet file** (no intermediate pandas step) that returns, for each integer `MedInc` bucket (`FLOOR(MedInc)`, i.e. \$10k-wide income bands), the row count, the mean and median `MedHouseVal` and the mean `AveRooms`, ordered by bucket. Then time reading only `["MedInc", "MedHouseVal"]` from CSV (`usecols=`) and from Parquet (`columns=`).

**Done when.**
- The DuckDB result matches the equivalent pandas `groupby` (`assert_frame_equal` after aligning names and order).
- You can explain the ordering of the two read timings using [columnar storage](#row-vs-columnar).
:::

:::{dropdown} Hint
Bucket with `np.floor(df["MedInc"])` on the pandas side. Avoid `ROUND` for this comparison: DuckDB and SQLite round halves *away from zero* (2.5 → 3), whereas pandas, NumPy and Polars round halves *to even* (2.5 → 2). California Housing has 54 `MedInc` values where the two rules disagree, so `ROUND` buckets would not match — a real-world reminder to check that two engines agree on semantics before comparing their outputs.
:::

:::{tip}
Wrap Polars/DuckDB imports in `try/except ImportError` so your notebook still runs on machines without those packages installed.
:::